cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/SsaInternals.qll

```diff
@@ -256,23 +257,63 @@ private predicate sourceVariableHasBaseAndIndex(SourceVariable v, BaseSourceVari
 }
 
 abstract class DefImpl extends DefOrUseImpl {
-  Operand address;
   int ind;
 
   bindingset[ind]
   DefImpl() { any() }
 
+  override int getIndirectionIndex() { result = ind }
+
+  override string toString() { result = "Def of " + this.getSourceVariable() }
+
   abstract int getIndirection();
 
+  abstract predicate isCertain();
+
   abstract Node0Impl getValue();
+}
 
-  abstract predicate isCertain();
+/** An initial definition of an `IRVariable`'s address. */
+private class DefAddressImpl extends DefImpl, TDefAddressImpl {
+  BaseIRVariable v;
 
-  Operand getAddressOperand() { result = address }
+  DefAddressImpl() {
+    this = TDefAddressImpl(v) and
+    ind = 0
+  }
 
-  override int getIndirectionIndex() { result = ind }
+  final override int getIndirection() { result = 0 }
 
-  override string toString() { result = "Def of " + this.getSourceVariable() }
+  final override predicate isCertain() { any() }
+
+  final override Node0Impl getValue() { none() }
+
+  final override predicate hasIndexInBlock(IRBlock block, int index) {
+    block = v.getIRVariable().getEnclosingIRFunction().getEntryBlock() and
+    index = 0
```

Comment:
Presumably this means multiple items can have index `0`. That's no problem?

Reply:
That's totally fine. We have the same situation for the initial definition of a global variable when entering a function: https://github.com/github/codeql/blob/main/cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/SsaInternals.qll#L576. In fact, we could probably have this predicate be identical to that one, but in practice it doesn't really matter since all we care about is that the SSA variable has a definition before its first use.